parsec/core/fs/workspacefs/versioning_helpers.py

```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
     destination: FsPath
 
 
-async def list_versions(
-    workspacefs, path: FsPath, skip_minimal_sync: bool = True
-) -> List[TimestampBoundedData]:
+class ManifestDataAndMutablePaths:
     """
-    Raises:
-        FSError
-        FSBackendOfflineError
-        FSWorkspaceInMaintenance
-        FSRemoteManifestNotFound
+    Contains a manifest and his known current, previous and next paths.
+    Unlike ManifestDataAndPaths, attributes are mutable.
+    """
+
+    def __init__(self, manifest, source_path=None, destination_path=None, current_path=None):
+        self.manifest = manifest
+        self.source = source_path
+        self.destination = destination_path
+        self.current = current_path
+
+
+class CacheEntry(NamedTuple):
+    """
+    Contains a manifest and the earliest and last timestamp for which its version has been returned
+    """
+
+    early: Pendulum
+    late: Pendulum
+    manifest: RemoteManifest
+
+
+class ManifestCache:
+    """
+    Caches manifest through their version number, and the timeframe for which they could be
+    obtained.
     """
-    # Could be optimized if we could use manifest.updated
-    manifest_cache = {}
-    versions_list_cache = {}
 
-    async def _load_manifest_or_cached(entry_id: EntryID, version=None, timestamp=None):
+    def __init__(self, remote_loader):
+        self._manifest_cache = {}
+        self._remote_loader = remote_loader
+
+    async def load(self, entry_id: EntryID, version=None, timestamp=None):
+        """
+        Raises:
+            FSError
+            FSBackendOfflineError
+            FSWorkspaceInMaintenance
+            FSRemoteManifestNotFound
+            FSBadEncryptionRevision
+            FSWorkspaceNoAccess
+        """
         try:
             if version:
-                return manifest_cache[entry_id][version][2]
+                return self._manifest_cache[entry_id][version].manifest
             if timestamp:
                 return next(
                     (
-                        t[2]
-                        for t in manifest_cache[entry_id].values()
-                        if t[0] and t[1] and t[0] <= timestamp <= t[1]
+                        t.manifest
+                        for t in self._manifest_cache[entry_id].values()
+                        if t.early and t.late and t.early <= timestamp <= t.late
                     )
                 )
+            return
         except (KeyError, StopIteration):
```

Comment:
The catching scope is a bit wide. Maybe manage the cache miss and the `StopIteration` separately?

Reply:
Done, and I splitted this code to another method to make the main one more readable!